dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/ShutdownServiceImpl.java

```diff
@@ -0,0 +1,192 @@
+                               final Workers workers) {
+        this.savedSessionStore = savedSessionStore;
+        this.savedJobPrioritiesStore = savedJobPrioritiesStore;
+        this.jobInterruptionStore = jobInterruptionStore;
+        this.settings = settings;
+        this.jobWorkers = jobWorkers;
+        this.workers = workers;
+    }
+
+    @Override
+    public void shutdown() {
+        closeApplication();
+    }
+
+    private void closeApplication() {
+        if (savedSessionStore != null) {
+            try {
+                SavedSessionStore.saveSavedSessionStore(savedSessionStore);
+            } catch (final IOException e) {
+                LOG.error("Failed to save session information to the local filesystem", e);
+            }
+        }
+        if (savedJobPrioritiesStore != null) {
+            try {
+                SavedJobPrioritiesStore.saveSavedJobPriorties(savedJobPrioritiesStore);
+            } catch (final IOException e) {
+                LOG.error("Failed to save job settings information to the local filesystem", e);
+            }
+        }
+
+        if (jobInterruptionStore != null) {
+            try {
+                JobInterruptionStore.saveJobInterruptionStore(jobInterruptionStore);
+            } catch (final Exception e) {
+                LOG.error("Failed to save job ids", e);
+            }
+        }
+
+        if (settings != null) {
+            try {
+                SettingsStore.saveSettingsStore(settings);
+            } catch (final IOException e) {
+                LOG.error("Failed to save settings information to the local filesystem", e);
+            }
+        }
+        if (jobWorkers.getTasks().size() != 0) {
+
+            final ImmutableList<Ds3JobTask> outstandingJobs = jobWorkers.getTasks().stream().collect(GuavaCollectors.immutableList());
+
+            final ShutdownTask task = new ShutdownTask(outstandingJobs);
+
+            workers.execute(task);
+            task.setOnSucceeded(event -> finalShutdown());
+        } else {
+            finalShutdown();
+        }
+    }
+
+    private void finalShutdown() {
+        workers.shutdown();
+        jobWorkers.shutdown();
+        jobWorkers.shutdownNow();
+        LOG.info("Finished shutting down");
+        Platform.exit();
+        System.exit(0);
+    }
+
+    private class ShutdownTask extends Task {
+
+        private final ImmutableList<Ds3JobTask> outstandingJobs;
+
+        public ShutdownTask(final ImmutableList<Ds3JobTask> outstandingJobs) {
+            this.outstandingJobs = outstandingJobs;
+        }
+
+
+        @Override
+        public Object call() {
+            outstandingJobs.forEach(job -> {
+                final CountDownLatch latch = new CountDownLatch(1);
+                try {
+
+                    String jobId = "";
+                    final Ds3Client ds3Client = job.getDs3Client();
+                    if (job instanceof Ds3PutJob) {
```

Comment:
Did we discuss Adding the `getJobId()` method to the parent class of Ds3PutJob?  If we did that this whole code segment would collapse down into a single block.